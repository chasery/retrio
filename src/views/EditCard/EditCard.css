.EditCard__wrapper {
  display: flex;
  justify-content: center;
}
