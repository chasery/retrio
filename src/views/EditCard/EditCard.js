import React, { useContext, useEffect, useState } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import BoardsApiService from '../../services/boards-api-service';
import CardsApiService from '../../services/cards-api-service';
import RetrioContext from '../../context/retrio-context';
import Header from '../../components/Header/Header';
import Form from '../../components/Form/Form';
import FormField from '../../components/FormField/FormField';
import Error from '../../components/Error/Error';
import './EditCard.css';

function EditCard(props) {
  const context = useContext(RetrioContext);
  const history = useHistory();
  const { boardId, cardId } = useParams();
  const [board, setBoard] = useState({});
  const [category, setCategory] = useState('');
  const [headline, setHeadline] = useState('');
  const [text, setText] = useState('');
  const [error, setError] = useState(null);

  useEffect(() => {
    async function initState() {
      try {
        let boardCall = await BoardsApiService.getBoard(boardId);
        let boardRes = await boardCall;

        setBoard(boardRes);

        let cardCall = await CardsApiService.getCard(cardId);
        let cardRes = await cardCall;

        setCategory(cardRes.category);
        setHeadline(cardRes.headline);
        setText(cardRes.text);
      } catch (error) {
        setError(error.error);
      }
    }

    initState();
  }, [boardId, cardId]);

  const handleEditCard = (e) => {
    e.preventDefault();

    CardsApiService.editCard(cardId, {
      board_id: boardId,
      category: category,
      headline: headline,
      text: text,
    })
      .then((res) => {
        history.push(`/boards/${boardId}`);
      })
      .catch((error) => setError(error.error));
  };

  const renderCategories = () => {
    const types = [];

    for (const [key, value] of Object.entries(context.cardCategories)) {
      types.push(
        <option key={key} value={`${key}`}>
          {value}
        </option>
      );
    }

    return types;
  };

  return (
    <>
      <Header />
      <main role='main'>
        <section className='EditCard'>
          <div className='EditCard__wrapper'>
            <Form id='EditCard' onSubmit={(e) => handleEditCard(e)}>
              <div className='Form__header'>
                <h2>Edit Card</h2>
              </div>
              <div className='Form__body'>
                <p>
                  Edit the following card in the{' '}
                  {board.name ? <strong>{board.name}</strong> : null}{' '}
                  retrospective.
                </p>
                <div className='FormField'>
                  <label className='FormField__label' htmlFor='category'>
                    Category<span className='FormField__required'>*</span>
                  </label>
                  <div className='FormField__select'>
                    <select
                      id='category'
                      required
                      value={category}
                      onChange={(e) => setCategory(e.target.value)}
                    >
                      <option value=''>Select Category</option>
                      {renderCategories()}
                    </select>
                  </div>
                </div>
                <FormField
                  id='headline'
                  label='Headline'
                  type='text'
                  isRequired={false}
                  onChange={(e) => setHeadline(e.target.value)}
                  value={headline}
                />
                <div className='FormField'>
                  <label className='FormField__label' htmlFor='text'>
                    Text<span className='FormField__required'>*</span>
                  </label>
                  <textarea
                    id='text'
                    className='FormField__textarea'
                    onChange={(e) => setText(e.target.value)}
                    rows='8'
                    value={text}
                  ></textarea>
                </div>
                {error ? <Error message={error} /> : null}
                <div className='Form__controls'>
                  <button
                    className='Form__button secondary'
                    type='button'
                    onClick={() => history.push(`/boards/${boardId}`)}
                  >
                    Cancel
                  </button>
                  <button className='Form__button primary' type='submit'>
                    Edit Card
                  </button>
                </div>
              </div>
            </Form>
          </div>
        </section>
      </main>
    </>
  );
}

export default EditCard;
