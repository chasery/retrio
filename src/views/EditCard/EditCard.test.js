import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import EditCard from './EditCard';

describe('EditCard component', () => {
  it('renders without error', () => {
    const wrapper = shallow(<EditCard />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
