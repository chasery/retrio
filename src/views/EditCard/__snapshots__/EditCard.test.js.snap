// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditCard component renders without error 1`] = `
<Fragment>
  <Header />
  <main
    role="main"
  >
    <section
      className="EditCard"
    >
      <div
        className="EditCard__wrapper"
      >
        <Form
          id="EditCard"
          onSubmit={[Function]}
        >
          <div
            className="Form__header"
          >
            <h2>
              Edit Card
            </h2>
          </div>
          <div
            className="Form__body"
          >
            <p>
              Edit the following card in the
               
               
              retrospective.
            </p>
            <div
              className="FormField"
            >
              <label
                className="FormField__label"
                htmlFor="category"
              >
                Category
                <span
                  className="FormField__required"
                >
                  *
                </span>
              </label>
              <div
                className="FormField__select"
              >
                <select
                  id="category"
                  onChange={[Function]}
                  required={true}
                  value=""
                >
                  <option
                    value=""
                  >
                    Select Category
                  </option>
                </select>
              </div>
            </div>
            <FormField
              id="headline"
              isRequired={false}
              label="Headline"
              onChange={[Function]}
              type="text"
              value=""
            />
            <div
              className="FormField"
            >
              <label
                className="FormField__label"
                htmlFor="text"
              >
                Text
                <span
                  className="FormField__required"
                >
                  *
                </span>
              </label>
              <textarea
                className="FormField__textarea"
                id="text"
                onChange={[Function]}
                rows="8"
                value=""
              />
            </div>
            <div
              className="Form__controls"
            >
              <button
                className="Form__button secondary"
                onClick={[Function]}
                type="button"
              >
                Cancel
              </button>
              <button
                className="Form__button primary"
                type="submit"
              >
                Edit Card
              </button>
            </div>
          </div>
        </Form>
      </div>
    </section>
  </main>
</Fragment>
`;
